---
title: "Mahad-Bihie-RNA-Seq-Analysis"
output:
  html_document:
    toc: true
    toc_float: true
    theme: united
    code_folding: hide
---


```{r libraries, message=FALSE}
#Required Libraries

#BiocManager::install("DESeq2")
library(DESeq2)
#BiocManager::install("EnsDb.Hsapiens.v86")
library(EnsDb.Hsapiens.v86)
#BiocManager::install("EnhancedVolcano")
library(EnhancedVolcano)
#BiocManager::install("pheatmap")
library(pheatmap)
#BiocManager::install("org.Hs.eg.db")
library(org.Hs.eg.db)
#BiocManager::install("KEGGREST")
library(KEGGREST)
#BiocManager::install("pathview")
library(pathview)
#BiocManager::install("clusterProfiler")
library(clusterProfiler)
#BiocManager::install("gage")
library(gage)
#BiocManager::install("gageData")
library(gageData)
#install.packages("tidyverse")
library(tidyverse)
#install.packages("recount")
library(recount)
#install.packages("RColorBrewer")
library(RColorBrewer)
#install.packages("DT")
library(DT)
#install.packages("enrichR")
library(enrichR)
#install.packages("kableExtra")
library(kableExtra)
```


## Construct the DESeq2 DataSet Object

I have created the DES object using the Ewing sarcoma RNA-seq data set provided. I have also made small changes to the DES object such as reducing the data to normalized values, removing low gene counts and removing the version number from the geneID.
```{r, message = FALSE, warning = FALSE}
#   class(rse): RangedSummarizedExperiment
rse <- readRDS("EwS.rds")

#remove the version number from the geneID
rownames(rse) <- gsub(rownames(rse),
                           pattern = "\\..+", replacement = "")

#make the dds
dds <- DESeqDataSet(rse, design = ~condition)

#set the factor level
dds$condition <- relevel(dds$condition, ref = "shCTR")

#creat DESeq 2 object
dds <- DESeq(dds)

#decrease the size of the DESeq object to make DESeq2 functions faster
dds <- estimateSizeFactors(dds)
idx <- rowSums( counts(dds, normalized=TRUE) >= 10 ) >= 3
dds <- dds[idx,]

#remove rows with low gene counts
keep <- as.data.frame(counts(dds)) %>%
  rowSums(counts(dds, normalized=TRUE)) >= 10
dds <- dds[keep,]
```

## 1. PCA Plot summarizing the sample-level variance within the data set.

A principal component analysis is used to emphasize the variation present in a dataset.This PCA plot displays the variation due to both the effect of the experiment and any batch effects as well. The control group and the experimental group are both clearly separated in this PCA with, with the first principal component dividing them with 98% of the variance.
```{r, message = FALSE}
#create rlog for PCA
rld <- rlog(dds)

#plot the PCA
plotadj <- plotPCA(rld)

#edit the plot height and width
plotadj + coord_fixed(ratio = 1,
                      xlim=c(-150, 150),
                      ylim=c(-100, 100))
```

## 2. MA Plot - Mean Expression VS Log 2 Fold Change

I first extracted the results and then shrunk the log fold change to remove any values where the fold change was exaggerated. The MA plot displays the relationship between the normalized mean expression and the log 2 fold change. The points shown at the edges of the y-axis (high absolute log fold change) are supposed to represent genes with high differential expression levels. As seen below, the only points with a high absolute log fold change also have high mean normalized counts. With the shrunken results, I was able to remove non-significant values with an extreme fold change to make a clearer graph that only presented differentially expressed values that are significant. There are also more significant genes with a positive log 2 fold change, indicating that there is a significant number of genes being upregulated. 

```{r, message = FALSE, warning = FALSE}
#extract the results
res <- results(dds,
               contrast = c("condition", 
                            "shCTR",      
                            "shEF1"),  
               alpha = 0.05)

#shrink data
resNorm <- lfcShrink(dds = dds,
                     res = res,
                     type = "normal",
                     coef = 2)

#MA Plot showing the relationship between mean count and log2 fold change.
plotMA(resNorm,
       main = "MA-plot of Normalized Ewing Sarcoma RNA-seq Data")
```

## 3. Table of Differentially Expressed Genes

I first used tidyverse code to align the data with the appropriate gene symbol. I then specified the leve of significance in the results and ordered the data according to the significance level. 
```{r, message = FALSE, warning = FALSE}

#make a dataframe of the results to view them
resdf <- as.data.frame(resNorm)

#extract gene symbol from EnsDb.Hsapiens.v86
ens2sym <- AnnotationDbi::select(EnsDb.Hsapiens.v86,
                                 keys = keys(EnsDb.Hsapiens.v86), 
                                 columns = c("SYMBOL"))

#join ens2sym to resdf by shared column (GENEID)
resdfsym <- resdf %>%
  rownames_to_column() %>%
  mutate(GENEID = gsub(rowname, pattern = "\\..+", replacement = "")) %>% 
  inner_join(y = ens2sym, by = "GENEID") %>% 
  dplyr::select(-rowname) %>%
  mutate(padj = case_when(padj == 0 ~ .Machine$double.xmin,
                          TRUE ~ padj)) #replacing 0s with minimum R value

#DF to specify level of signficance
DEgenes <- resdfsym %>%
  arrange(padj) %>%
  dplyr::filter(padj < 0.05 & log2FoldChange > 2) %>%  
  #slice_head(n=10) %>%
  select("SYMBOL", "GENEID", "padj", "baseMean", "log2FoldChange", "lfcSE", "stat", "pvalue") 

#data table of DE genes
datatable(DEgenes,
          class = 'cell-border stripe',
          caption = htmltools::tags$caption(
          style = 'caption-side: bottom; text-align: center;',
                  'Table 1: ',
                  width = 3,
          htmltools::em('Differentially Expressed Genes.')),
          filter = 'top',
          extensions = 'FixedColumns',
          options = list(
              pageLength = 5,
              autoWidth = TRUE
              ,
              scrollX = TRUE
                        )
          )
```

## 4. Volcano Plot of all Differentially Expressed Genes


Volcano plots display the statistical significance of the difference relative to the magnitude of difference for every single gene in the comparison. This volcano plot presents a wide dispersion between the control and experimental group, which implies that there is a strong difference in gene expression. For example, COL11A1 is shown to have a small p-value and is therefore considered statistically differentially expressed. The are also many genes far from 0 which shows that are large differences in mean expression levels. Question 3 also supports the significance of COLL1A1.
```{r, message = FALSE, warning = FALSE, fig.height = 10, fig.width = 10}
#create the volcano plot
EnhancedVolcano(resdfsym, lab = resdfsym$SYMBOL, pCutoff = 0.05,
                FCcutoff = 2,
                x = "log2FoldChange",
                y = "padj",
                #xlim = c(-30, 30),
                title = "EWSR1-FLI1 Suppressed Ewing Sarcoma DEGs")
```

## 5. Heatmap of Top 10 Over Expressed & Under expressed DEGs
The heatmap below shows a clear distinction between the untreated and treated samples in terms of gene expression. Under the treated columns, you can see which genes are over-expressed and under-expressed as they are coloured according to the gradient on the right. The untreated samples are down regulated for the top 10 over-expressed genes shown, with mixed results for the bottom 10 under expressed genes. 
```{r, message = FALSE}
#filter df to significant DEGs that are both over expressed and underexpressed
orderedSig <-resdfsym %>%
  arrange(padj) %>%
  dplyr::filter(padj < 0.05 & log2FoldChange > 2 | log2FoldChange < -2) 

#create matrix for heat map
id1 <- orderedSig$GENEID 
id2 <- orderedSig$SYMBOL 
mat <- assay(rld) #extract mtx of nrmlzd counts
DE <- mat[id1,]
rownames(DE) <- id2 #rmv ensemble iDs from row name

#transform matrix to df to filter to top 10 DEGs
#overexpressed
overDE <- as.data.frame(DE) %>% #filter to top 10 overexpressed genes
  slice_head(n=10)
overDE <- as.matrix(overDE) #make it a matrix for pheatmap
#underexpressed
underDE <- as.data.frame(DE) %>% #filter to top 10 underrexpressed genes
  slice_tail(n=10)
underDE <- as.matrix(underDE) #make it a matrix for pheatmap


DE <- rbind(overDE, underDE) #bind both matrices into 1
annotation <- as.data.frame(colData(rld)[, c("run","condition")]) #add annotation onto map

#heatmap
pheatmap(DE,
         scale = "row", 
         clustering_distance_rows = "correlation", 
         cluster_rows=F, 
         cluster_cols=F,
         annotation_col = annotation,
         show_colnames = FALSE, # removes coloumn names from run at the btm
         annotation_names_col = FALSE, #removes names beside the colours at the top
         main="Top 10 Differentially Expressed genes"
)
```

## 6. Enrichment Analysis of the top over- and under-expressed KEGG pathways. 
I used the gage() function to generate the top over-expressed and under-expressed KEGG pathways, using the results of the DESeq2 object. I then created a data table of the pathways and test statistics. I also included the enrichment scores from the gseKEGG() function. Lastly, I added the the top over-expressed and under-expressed KEGG pathway below.  
```{r, results = FALSE, message = FALSE, warning = FALSE}
#reorder these results by p-value and call summary() on the results object
res = res[order(res$pvalue),]
summary(res)
res$symbol = mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first")
res$entrez = mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")
res$name =   mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="GENENAME",
                     keytype="ENSEMBL",
                     multiVals="first")
#setup the KEGG data-sets we need.
data(kegg.sets.hs)
data(sigmet.idx.hs)
#remove unnecessary pathway defintions
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]
#creating vector of FCs with Entrez IDs as the names for the gage() function
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
#Get the results for the pathway analysis
keggres = gage(foldchanges, gsets=kegg.sets.hs, same.dir=TRUE)
```
### A datatable of the Top Over-Expressed & Under-Expressed KEGG Pathways.
```{r}
#join the over-expressed and under-expressed pathways
keggtable <- c(rownames(keggres$greater),rev(rownames(keggres$less)))

#edit the dataframe
keggtable <- as.data.frame(keggtable) 

keggtable <- keggtable %>%
  separate(keggtable,
           sep = 9,
           into=c("pathid","pathway")) %>% 
  mutate(pathid = substring(pathid, 4)) %>%
  select(pathway,pathid) %>% 
  distinct()

#format p value for upregulated and downregulated pathways
grtr <- as.data.frame(keggres$greater) %>%
  rownames_to_column() %>%
  rename(pathway = rowname) %>%
  mutate(pathway = substring(pathway, 10)) %>% na.omit() %>% distinct()

#join dataframes
KT <- grtr %>%
  left_join(keggtable, by="pathway") %>%
  select(pathway, pathid, p.val, stat.mean) %>%
  distinct()

#present the results in a data table
datatable(KT, 
          options=list(pageLength=5),
          caption = htmltools::tags$caption(
          style = 'caption-side: bottom; text-align: center;',
                  'Table 2: ',
                  width = 3,
          htmltools::em('Top Over-Expressed & Under-Expressed KEGG Pathways.')
                                           )
         )

```
### The Enrichment Scores from gseKEGG() Function.
```{r, message = FALSE, warning = FALSE}
#create a df of the results from the modified resNorm
resdf2 <- as.data.frame(resNorm)

# we want the log2 fold change 
original_gene_list <- resdf2$log2FoldChange

# name the vector
names(original_gene_list) <- rownames(resdf2)

# omit any NA values 
original_gene_list <- na.omit(original_gene_list)

# sort the list in decreasing order (required for clusterProfiler)
original_gene_list = sort(original_gene_list, decreasing = TRUE)

# Convert gene IDs for gseKEGG function
    #lose some genes here because not all IDs will be converted
ids <- bitr(geneID = names(original_gene_list), 
          fromType = "ENSEMBL", 
          toType = "ENTREZID", 
          OrgDb = org.Hs.eg.db)

# remove duplicate IDS (here I use "ENSEMBL", but it should be whatever was selected as keyType)
dedup_ids = ids[!duplicated(ids[c("ENSEMBL")]),]

df2 = resdf2[rownames(resdf2) %in% dedup_ids$ENSEMBL,]

df2$Y = dedup_ids$ENTREZID

# Create a vector of the gene universe
kegg_gene_list <- df2$log2FoldChange

# Name vector with ENTREZ ids
names(kegg_gene_list) <- df2$Y

# omit any NA values 
kegg_gene_list<-na.omit(kegg_gene_list)

# sort the list in decreasing order (required for clusterProfiler)
kegg_gene_list = sort(kegg_gene_list, decreasing = TRUE)

kegg_organism = "hsa"
kk2 <- gseKEGG(geneList     = kegg_gene_list,
               organism     = kegg_organism,
               #nPerm        = 10000,
               #minGSSize    = 3,
               #maxGSSize    = 800,
               pvalueCutoff = 0.05,
               pAdjustMethod = "none",
               keyType       = "ncbi-geneid")


#display results
enrichkegg <- kk2@result
enrichkegg <- enrichkegg %>%
  rownames_to_column() %>%
   rename(pathid = rowname) %>%
   select(pathid, Description, enrichmentScore) 

#edit description 
enrichkegg$Description = substr(enrichkegg$Description, 1, nchar(enrichkegg$Description)-22)

#data table for enrichment scores
datatable(enrichkegg,
          class = 'cell-border stripe',
          caption = htmltools::tags$caption(
          style = 'caption-side: bottom; text-align: center;',
                  'Table 3: ',
                  width = 3,
          htmltools::em('Enrichment Scores of KEGG Pathways.')),
          filter = 'top',
          extensions = 'FixedColumns',
          options = list(
              pageLength = 5,
              autoWidth = TRUE
              ,
              scrollX = TRUE
                        )
          )
```



### The Top Under-Expressed & Over-Expressed KEGG Pathways.
```{r, results = FALSE, message = FALSE, warning = FALSE}
#top overexpressed pathway
okeggrespathways <- rownames(keggres$greater)[1]

#top underexpressed pathway
ukeggrespathways <- rownames(keggres$less)[1]

# Extract the IDs part of each string
okeggresids = substr(okeggrespathways, start=1, stop=8)
ukeggresids = substr(ukeggrespathways, start=1, stop=8)

#pass these IDs in keggkresids to the pathview() function to draw top pathways
opathres <- pathview(gene.data=foldchanges, pathway.id=okeggresids, species="hsa")
upathres <- pathview(gene.data=foldchanges, pathway.id=ukeggresids, species="hsa")
```


**The Top Over-Expressed KEGG Pathway: NUCLEOCYTOPLASMIC TRANSPORT**
```{r out.width = "100%"}
knitr::include_graphics(paste(okeggresids,"pathview.png",sep = "."), error = FALSE)
```

**The Top Under-Expressed KEGG Pathway: ECM-RECEPTOR INTERACTION**
```{r out.width = "100%"}
knitr::include_graphics(paste(ukeggresids,"pathview.png",sep = "."), error = FALSE)
```









