---
title: "Mahad-Bihie-RNA-Seq-Analysis"
output:
  html_document:
    toc: true
    toc_float: true
    theme: united
---

## Required Libraries
```{r libraries, message=FALSE}
#BiocManager::install("DESeq2")
library(DESeq2)
#BiocManager::install("EnsDb.Hsapiens.v86")
library(EnsDb.Hsapiens.v86)
#BiocManager::install("EnhancedVolcano")
library(EnhancedVolcano)
#BiocManager::install("pheatmap")
library(pheatmap)
#BiocManager::install("org.Hs.eg.db")
library(org.Hs.eg.db)
#BiocManager::install("KEGGREST")
library(KEGGREST)
#BiocManager::install("pathview")
library(pathview)
#BiocManager::install("gage")
library(gage)
#BiocManager::install("gageData")
library(gageData)
#install.packages("tidyverse")
library(tidyverse)
#install.packages("recount")
library(recount)
#install.packages("RColorBrewer")
library(RColorBrewer)
#install.packages("DT")
library(DT)
#install.packages("enrichR")
library(enrichR)
#install.packages("kableExtra")
library(kableExtra)
```


## Construct the DESEQDataSet Object
```{r, message = FALSE, warning = FALSE}
#read the data into R
#   class(rse): RangedSummarizedExperiment
rse <- readRDS("EwS.rds")

#remove the version number from the geneID
rownames(rse) <- gsub(rownames(rse),
                           pattern = "\\..+", replacement = "")

#make the dds
dds <- DESeqDataSet(rse, design = ~condition)

#set the factor level
dds$condition <- relevel(dds$condition, ref = "shCTR")

#creat DESeq 2 object
dds <- DESeq(dds)

#decrease the size of the DESeq object to make DESeq2 functions faster
dds <- estimateSizeFactors(dds)
idx <- rowSums( counts(dds, normalized=TRUE) >= 10 ) >= 3
dds <- dds[idx,]

#remove rows with low gene counts
keep <- as.data.frame(counts(dds)) %>%
  rowSums(counts(dds, normalized=TRUE)) >= 10
dds <- dds[keep,]
```

## 1. PCA Plot summarizing the sample-level variance within the data set.
```{r, message = FALSE}
#create rlog for PCA
rld <- rlog(dds)

#plot the PCA
plotPCA(rld)
```

## 2. MA Plot - Relationship between Mean Mount and Log2 Fold Change.
```{r, message = FALSE, warning = FALSE}
#extract the results
res <- results(dds,
               contrast = c("condition", 
                            "shCTR",      
                            "shEF1"),  
               alpha = 0.05)

#shrink data
resNorm <- lfcShrink(dds = dds,
                     res = res,
                     type = "normal",
                     coef = 2)

#MA Plot showing the relationship between mean count and log2 fold change.
plotMA(resNorm,
       main = "MA-plot of Normalized Ewing Sarcoma RNA-seq Data")
```

## 3. Volcano Plot of all Differentially Expressed Genes
```{r, message = FALSE, warning = FALSE}

#make a dataframe of the results to view them
resdf <- as.data.frame(resNorm)

#extract gene symbol from EnsDb.Hsapiens.v86
ens2sym <- AnnotationDbi::select(EnsDb.Hsapiens.v86,
                                 keys = keys(EnsDb.Hsapiens.v86), 
                                 columns = c("SYMBOL"))

#join ens2sym to resdf by shared column (GENEID)
resdfsym <- resdf %>%
  rownames_to_column() %>%
  mutate(GENEID = gsub(rowname, pattern = "\\..+", replacement = "")) %>% 
  inner_join(y = ens2sym, by = "GENEID") %>% 
  dplyr::select(-rowname) %>%
  mutate(padj = case_when(padj == 0 ~ .Machine$double.xmin,
                          TRUE ~ padj)) #replacing 0s with minimum R value

#create the volcano plot
EnhancedVolcano(resdfsym, lab = resdfsym$SYMBOL, pCutoff = 1e-100,
                FCcutoff = 3,
                x = "log2FoldChange",
                y = "padj",
                title = "EWSR1-FLI1 Suppressed Ewing Sarcoma DEGs")
```

## 4. Table of Differentially Expressed Genes
```{r, message = FALSE, warning = FALSE, out.width="30%"}
#DF to specify level of signficance
DEgenes <- resdfsym %>%
  arrange(padj) %>%
  dplyr::filter(padj < 0.05 & log2FoldChange > 2) %>%  
  #slice_head(n=10) %>%
  select("SYMBOL", "GENEID", "padj", "baseMean", "log2FoldChange", "lfcSE", "stat", "pvalue") 

#data table of DE genes
datatable(DEgenes,
          class = 'cell-border stripe',
          caption = htmltools::tags$caption(
          style = 'caption-side: bottom; text-align: center;',
                  'Table 1: ',
                  width = 3,
          htmltools::em('Differentially Expressed Genes.')),
          filter = 'top',
          extensions = 'FixedColumns',
          options = list(
              pageLength = 5,
              autoWidth = TRUE
              ,
              scrollX = TRUE
                        )
          )

```
## 5. Heatmap of Top 10 Over Expressed & Under expressed DEGs
```{r, message = FALSE}
#filter df to significant DEGs that are both over expressed and underexpressed
orderedSig <-resdfsym %>%
  arrange(padj) %>%
  dplyr::filter(padj < 0.05 & log2FoldChange > 2 | log2FoldChange < -2) 

#create matrix for heat map
id1 <- orderedSig$GENEID 
id2 <- orderedSig$SYMBOL 
mat <- assay(rld) #extract mtx of nrmlzd counts
DE <- mat[id1,]
rownames(DE) <- id2 #rmv ensemble iDs from row name

#transform matrix to df to filter to top 10 DEGs
#overexpressed
overDE <- as.data.frame(DE) %>% #filter to top 10 overexpressed genes
  slice_head(n=10)
overDE <- as.matrix(overDE) #make it a matrix for pheatmap
#underexpressed
underDE <- as.data.frame(DE) %>% #filter to top 10 underrexpressed genes
  slice_tail(n=10)
underDE <- as.matrix(underDE) #make it a matrix for pheatmap


DE <- rbind(overDE, underDE) #bind both matrices into 1
annotation <- as.data.frame(colData(rld)[, c("run","condition")]) #add annotation onto map

#heatmap
pheatmap(DE,
         scale = "row", 
         clustering_distance_rows = "correlation", 
         annotation_col = annotation, 
         main="Top 10 Differentially Expressed genes"
)
```

## 6. Enrichment Analysis of the top over- and under-expressed KEGG pathways. 
```{r, results = FALSE, message = FALSE, warning = FALSE}
#, echo=FALSE, message = FALSE, warning = FALSE
#reorder these results by p-value and call summary() on the results object
res = res[order(res$pvalue),]
summary(res)

res$symbol = mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first")

res$entrez = mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")

res$name =   mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="GENENAME",
                     keytype="ENSEMBL",
                     multiVals="first")

#setup the KEGG data-sets we need.
data(kegg.sets.hs)
data(sigmet.idx.hs)

#remove unnecessary pathway defintions
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

#creating vector of FCs with Entrez IDs as the names for the gage() function
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez

#Get the results for the pathway analysis
keggres = gage(foldchanges, gsets=kegg.sets.hs, same.dir=TRUE)
```
### A datatable of Top Over-Expressed & Under-Expressed KEGG Pathways.
```{r}
#join the over expressed and under expressed pathways
keggtable <- c(rownames(keggres$greater),rev(rownames(keggres$less)))

#edit the dataframe
keggtable <- as.data.frame(keggtable) 
keggtable <- keggtable %>%
  separate(keggtable,
           sep = 9,
           into=c("pathid","pathway")) %>% 
  select(pathway,pathid)
  
#present the results in a data table
datatable(keggtable, 
          options=list(pageLength=5),
          caption = htmltools::tags$caption(
          style = 'caption-side: bottom; text-align: center;',
                  'Table 2: ',
                  width = 3,
          htmltools::em('Top Over-Expressed & Under-Expressed KEGG Pathways.')
                                           )
         )
```

### The Top Under-Expressed & Over-Expressed KEGG Pathways.
```{r, results = FALSE, message = FALSE, warning = FALSE}
#top overexpressed pathway
okeggrespathways <- rownames(keggres$greater)[1]

#top underexpressed pathway
ukeggrespathways <- rownames(keggres$less)[1]

# Extract the IDs part of each string
okeggresids = substr(okeggrespathways, start=1, stop=8)
ukeggresids = substr(ukeggrespathways, start=1, stop=8)

#pass these IDs in keggkresids to the pathview() function to draw top pathways
opathres <- pathview(gene.data=foldchanges, pathway.id=okeggresids, species="hsa")
upathres <- pathview(gene.data=foldchanges, pathway.id=ukeggresids, species="hsa")
```


**The Top Over-Expressed KEGG Pathway: NUCLEOCYTOPLASMIC TRANSPORT**
```{r out.width = "100%"}
knitr::include_graphics(paste(okeggresids,"pathview.png",sep = "."), error = FALSE)
```

**The Top Under-Expressed KEGG Pathway: ECM-RECEPTOR INTERACTION**
```{r out.width = "100%"}
knitr::include_graphics(paste(ukeggresids,"pathview.png",sep = "."), error = FALSE)
```









